### ISSUE TITLE
ActionCable requests High Response Time

### ISSUE THREAD

**Issue by cesar3030** [rails/rails#35590]
High Response Time on `/cable` endpoint when using ActionCable with Puma server.

Here is a screenshot of the requests to my Puma server deployed on Heroku (Professional Dyno with 512Mo RAM). Most of the requests to `/cable` endpoint are taking a very long time.

Am I the only one experiencing this?

![Screen Shot 2019-03-12 at 12 05 29 PM](https://user-images.githubusercontent.com/3688337/54235714-43aca480-44e8-11e9-8278-806b0e7178df.png)

### Expected behavior

`/cable` requests should be processed in a few ms.

### Actual behavior

`/cable` requests are taking ages to be processed.

### System configuration

**Ruby version**: 2.5.1
**Rails version**: 5.2.2
**Puma version**: 3.12.0

**Comment by jeremy** [comment]
Hey @cesar3030, you'll need to dig in and do some debugging and diagnosis to discover what's happening. As it stands, this is a deployment issue that you'll need to troubleshoot further. If you come across an issue with Action Cable implementation, please do raise it here.

**Comment by houndtt** [comment]
Hey @cesar3030, I've noticed this with my application on Heroku as well. Similar Puma version with older Ruby but same behaviour intermittently. I'm wondering if this is supposed to behave this way and I should just exclude the path from response time checks as suggested [here](https://stackoverflow.com/questions/45171520/actioncable-server-response-times-on-heroku).

Any update on the issue in your setup? Did you get a resolution to the issue?

**Comment by cesar3030** [comment]
Hi @houndtt, 
Unfortunately I can't help you since we ended up removing the ActionCable part of our application but I can give you an answer I got from the Heroku support team. I hope this will help...

Hey Cesar,

Thanks for reaching out with your questions about ActionCable performance on Heroku.

I actually haven't seen a ton of use cases around ActionCable performance on Heroku. I do know some folks have started looking into [AnyCable](https://github.com/anycable/anycable) - which touts lower memory consumption / performance because it leverages Go. However, I still haven't taken time to compare the two libraries.

The best explanation I've found so far for long ActionCable response times is that a Rails server (in your case Puma) has to traverse the entire list of connections to process data from ActionCable. I'm not entirely sure if this causes a burden on the `/cable` endpoint, but it could explain the high response time. There is some [current work](https://github.com/puma/puma/pull/1728) being done on Puma to remedy this though.

In short, I'm not entirely sure why exactly the performance is so expensive for ActionCable. It's ultimately going to be something best taken up with the folks for maintain it around. If you don't mind, take some time to open an issue with them and see if there's a reason for such expensive performance on a 512MB machine. I myself am very interested to see what you can dig up from that.

Best,

**Comment by j4rs** [comment]
Stopping over here to let you know that this hasn't stopped even after updating to the latest version of Puma that includes PR https://github.com/puma/puma/pull/1728 🤷‍♂.

**Comment by birthdaycorp** [comment]
this is still happening with me. i'm checking out AnyCable.

**Comment by jessehanley** [comment]
FYI I know this is closed but this is still happening to me over here. Will begin doing some research now.
